pkg/storage/endpoint/gc_states.go

```diff
@@ -0,0 +1,524 @@
+// designed for arbitrary use, but restricted for upper layers of GC state management modules only.
+//
+// Explanations of concepts mentioned in this file (here the term `snapshots` means snapshots of TiKV's MVCC data,
+// represented by a timestamp):
+//
+//   - GC Safe Point: A timestamp, the snapshots before which can be safely discarded by GC. Written by the GCWorker
+//     to control the GC procedure.
+//   - Txn Safe Point / Transaction Safe Point: A timestamp, the snapshots equal to or after which can be safely read.
+//     Written by the GCWorker to control the GC procedure.
+//   - GC Barriers: Blocks GC from advancing the txn safe point over some specific timestamps (the barrierTS of these
+//     barriers), ensures snapshots equal to or after which to be safe to read. GC barriers can be set by any components
+//     in the cluster.
+//   - Service Safe Points / Service GC Safe Points: Another mechanism that has the same purpose as GC barriers, but
+//     is planned to be deprecated in favor of GC barriers. However, in order to keep the backward compatibility of the
+//     persistent data, the data structure of service safe points is still used internally to represent GC barriers.
+//     Service safe points can also be set by any components in the cluster.
+//
+// GC management may differ between different keyspaces. There are two kinds of GC management, each of which has
+// different paths to write its metadata in etcd:
+//
+//   - Keyspace-level: A keyspace manages its GC by itself, and have independent GC states from other keyspaces.
+//   - Unified: Keyspaces not configured to use keyspace-level GC are running unified GC. The NullKeyspace (which is
+//     used when a TiDB node are not configured to use any keyspace) is always running unified GC. For all keyspaces
+//     running unified GC, the GC states are shared and uniformly managed by the NullKeyspace.
+
+// ServiceSafePoint is the service safe point for a specific service.
+// NOTE:
+//   - This type is exported by HTTP API. Please pay more attention when modifying it.
+//   - This type is in sync with `client/http/types.go`.
+//   - ServiceSafePoint is also directly used for storing GC barriers in order to make GC barriers in new versions
+//     backward-compatible with service safe points in old versions.
+//   - The concept "service safe point" is planned to be deprecated once GC barriers are ready to use.
+type ServiceSafePoint struct {
+	ServiceID string
+	ExpiredAt int64
+	SafePoint uint64
+
+	// Note than when marshalled into JSON, omitting KeyspaceID stands for the NullKeyspace (0xffffffff),
+	// rather than KeyspaceID = 0 which is the ID of the default keyspace.
+	// Special marshalling / unmarshalling methods are given for handling this field in a non-default way.
+	//
+	// The purpose is to make the code (for keyspaced and non-keyspaced/global GC) unified while keeping the
```

Comment:
How about we consistently use 'unified' in the comments to represent the GC of Null keyspace, and use 'Null keyspace' to represent the situation of 'unconfigured keyspace'? Let's avoid using terms like 'non-keyspaced' and 'global GC' or other expressions. This will help standardize the definitions of these concepts.

Reply:
It seems some comments are missed when I update it. I'll check it again.